src/Util/Lock.php

```diff
@@ -21,191 +19,79 @@
  */
 class Lock
 {
-	private static $semaphore = [];
-
 	/**
-	 * @brief Check for memcache and open a connection if configured
-	 *
-	 * @return object|boolean The memcache object - or "false" if not successful
+	 * @var Lock\ILockDriver;
 	 */
-	private static function connectMemcache()
-	{
-		if (!function_exists('memcache_connect')) {
-			return false;
-		}
-
-		if (!Config::get('system', 'memcache')) {
-			return false;
-		}
+	static $driver = null;
 
-		$memcache_host = Config::get('system', 'memcache_host', '127.0.0.1');
-		$memcache_port = Config::get('system', 'memcache_port', 11211);
-
-		$memcache = new Memcache;
-
-		if (!$memcache->connect($memcache_host, $memcache_port)) {
-			return false;
+	public static function init()
+	{
+		switch(Config::get('system', 'lock_driver', 'default')) {
+			case 'memcache':
+				self::$driver = new Lock\MemcacheLockDriver();
+				break;
+			case 'database':
+				self::$driver = new Lock\DatabaseLockDriver();
+				break;
+			case 'semaphore':
+				self::$driver = new Lock\SemaphoreLockDriver();
+				break;
+			default:
+				// Determine what's the best/fastest locking driver (default behavior in Friendica)
+				if (function_exists('sem_get') && version_compare(PHP_VERSION, '5.6.1', '>=')) {
+					self::$driver = new Lock\SemaphoreLockDriver();
+				} elseif (Config::get('system', 'cache_driver', 'database') == 'memcache') {
```

Comment:
There also are a `memcached` and a `redis` cache driver, so this check should just exclude `database`.

Reply:
You just just have to be aware when you add any new cache driver that it will automatically be used for locks. Is that suitable for us?

Reply:
I think so, cache is supposed to be temporary storage, so it can be used as a lock mechanism.

Reply:
I just thought that it might not be suitable for you in your multi-host endeavor. If we have a cache driver local to each host, then locks wouldn’t be shared.

This could happen with APCu for example, if we decide to support it.

Reply:
Hm.. That's why I think it would he better to explicit define the lock driver..

So you can use APCu / memcache for local caches but use redis / mencached too. I know nextcloud f.e. is working like that

Reply:
I see two solutions:
- Add a multi-host switch to avoid single host cache drivers.
- Add a lock driver for each cache driver and bypass the Cache abstraction entirely.

Reply:
I take the second one :) .. I feel it's more transparent

Reply:
Good for you, it is the righest path! The other feels hacky indeed.